File: src/main/java/br/com/shoppinglist/shopping_list/identity/infrastructure/adapter/in/web/AuthController.java
package br.com.shoppinglist.shopping_list.identity.infrastructure.adapter.in.web;

import br.com.shoppinglist.shopping_list.identity.application.dto.AuthTokenResult;
import br.com.shoppinglist.shopping_list.identity.application.dto.ConfirmRegistrationCommand;
import br.com.shoppinglist.shopping_list.identity.application.dto.LoginUserCommand;
import br.com.shoppinglist.shopping_list.identity.application.dto.RequestRegistrationCodeCommand;
import br.com.shoppinglist.shopping_list.identity.application.port.in.ConfirmRegistrationUseCase;
import br.com.shoppinglist.shopping_list.identity.application.port.in.LoginUserUseCase;
import br.com.shoppinglist.shopping_list.identity.application.port.in.RequestRegistrationCodeUseCase;
import br.com.shoppinglist.shopping_list.identity.infrastructure.adapter.in.web.dto.AuthResponse;
import br.com.shoppinglist.shopping_list.identity.infrastructure.adapter.in.web.dto.ConfirmRegistrationRequest;
import br.com.shoppinglist.shopping_list.identity.infrastructure.adapter.in.web.dto.LoginRequest;
import br.com.shoppinglist.shopping_list.identity.infrastructure.adapter.in.web.dto.MessageResponse;
import br.com.shoppinglist.shopping_list.identity.infrastructure.adapter.in.web.dto.RequestRegistrationCodeRequest;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/auth")
public class AuthController {

	private final RequestRegistrationCodeUseCase requestRegistrationCodeUseCase;
	private final ConfirmRegistrationUseCase confirmRegistrationUseCase;
	private final LoginUserUseCase loginUserUseCase;

	public AuthController(
		RequestRegistrationCodeUseCase requestRegistrationCodeUseCase,
		ConfirmRegistrationUseCase confirmRegistrationUseCase,
		LoginUserUseCase loginUserUseCase
	) {
		this.requestRegistrationCodeUseCase = requestRegistrationCodeUseCase;
		this.confirmRegistrationUseCase = confirmRegistrationUseCase;
		this.loginUserUseCase = loginUserUseCase;
	}

	@PostMapping("/register/request-code")
	@ResponseStatus(HttpStatus.ACCEPTED)
	public MessageResponse requestRegistrationCode(@Valid @RequestBody RequestRegistrationCodeRequest request) {
		var result = requestRegistrationCodeUseCase.execute(new RequestRegistrationCodeCommand(request.email()));
		return new MessageResponse(result.message());
	}

	@PostMapping("/register/confirm")
	@ResponseStatus(HttpStatus.CREATED)
	public AuthResponse confirmRegistration(@Valid @RequestBody ConfirmRegistrationRequest request) {
		AuthTokenResult result = confirmRegistrationUseCase.execute(
			new ConfirmRegistrationCommand(
				request.email(),
				request.code(),
				request.name(),
				request.password()
			)
		);
		return toResponse(result);
	}

	@PostMapping("/login")
	public AuthResponse login(@Valid @RequestBody LoginRequest request) {
		AuthTokenResult result = loginUserUseCase.execute(
			new LoginUserCommand(request.email(), request.password())
		);
		return toResponse(result);
	}

	private AuthResponse toResponse(AuthTokenResult result) {
		return new AuthResponse(result.accessToken(), result.tokenType(), result.expiresInSeconds());
	}
}

File: src/auth/auth.controller.ts
import {
  Body,
  Controller,
  HttpCode,
  HttpStatus,
  Post,
  UseFilters,
} from '@nestjs/common';
import { AuthExceptionFilter } from './auth.exception-filter';
import { AuthService } from './auth.service';
import { ConfirmRegistrationDto } from './dto/confirm-registration.dto';
import { LoginDto } from './dto/login.dto';
import { RequestRegistrationCodeDto } from './dto/request-registration-code.dto';

@Controller('auth')
@UseFilters(AuthExceptionFilter)
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('register/request-code')
  @HttpCode(HttpStatus.ACCEPTED)
  requestRegistrationCode(@Body() dto: RequestRegistrationCodeDto) {
    return this.authService.requestRegistrationCode(dto);
  }

  @Post('register/confirm')
  @HttpCode(HttpStatus.CREATED)
  confirmRegistration(@Body() dto: ConfirmRegistrationDto) {
    return this.authService.confirmRegistration(dto);
  }

  @Post('login')
  @HttpCode(HttpStatus.OK)
  login(@Body() dto: LoginDto) {
    return this.authService.login(dto);
  }
}
